File: compat/suites/java-sdk/src/test/java/io/overcast/compat/scenario/NowFixtureTest.java
package io.overcast.compat.scenario;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.junit.jupiter.api.Assumptions;
import org.junit.jupiter.api.DynamicNode;
import org.junit.jupiter.api.DynamicTest;
import org.junit.jupiter.api.TestFactory;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.function.LongSupplier;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.junit.jupiter.api.Assertions.fail;

/**
 * The shared {@code $now} conformance fixture, {@code compat/model/testdata/now}.
 *
 * <p>{@code $now} is the client's clock when a call is made, in epoch
 * milliseconds, plus an offset ({@code compat/model/README.md} § Values). The
 * emitted source hands this runtime {@code Values.now(unit, offsetMillis)} —
 * never the JSON — so the fixture's {@code invalid} spellings are the
 * generator's to refuse, and this runs the rest: each valid spelling binds to
 * the fixture's value into the {@code Long} a long member is, every invalid
 * argument is refused naming the member, and the clock is read once per call
 * however many {@code $now}s the call holds.
 *
 * <p>Found and required exactly as {@link BlobFixtureTest} finds its fixture.
 */
class NowFixtureTest {

    private static final String REQUIRED_ENV_VAR = "OVERCAST_COMPAT_FIXTURES_REQUIRED";

    /** A clock that moves on by {@code tick} every time it is read. */
    private static LongSupplier ticking(long start, long tick) {
        long[] next = {start};
        return () -> {
            long reading = next[0];
            next[0] += tick;
            return reading;
        };
    }

    private static Binder binder(LongSupplier clock) {
        return new Binder("oc-test", "logs-events", new ContextBag(), clock);
    }

    @TestFactory
    List<DynamicNode> sharedNowFixture() throws IOException {
        File file = fixtureFile();
        if (file == null) {
            if ("1".equals(System.getenv(REQUIRED_ENV_VAR))) {
                fail(REQUIRED_ENV_VAR + "=1 but no compat/model/testdata/now/now.json found above the"
                        + " working directory — this suite's fixture test must run from a full checkout");
            }
            return List.of(DynamicTest.dynamicTest("shared $now fixture", () -> Assumptions.abort(
                    "no compat/model/testdata/now above the working directory;"
                            + " run `mvn -B test` from compat/suites/java-sdk to check it")));
        }
        JsonNode root = new ObjectMapper().readTree(file);
        for (Iterator<String> it = root.fieldNames(); it.hasNext(); ) {
            String key = it.next();
            assertTrue(List.of("$comment", "instant", "valid", "invalid", "invalidArguments", "call").contains(key),
                    "unknown key " + key + " in " + file);
        }
        assertTrue(root.get("valid").size() > 0 && root.get("invalid").size() > 0
                        && root.get("invalidArguments").size() > 0,
                "the $now fixture may not be skipped by emptying it");
        long instant = root.get("instant").asLong();
        JsonNode call = root.get("call");
        long tick = call.get("tickMillis").asLong();

        List<DynamicNode> out = new ArrayList<>();
        for (JsonNode c : root.get("valid")) {
            String unit = c.get("now").get("unit").asText();
            long offset = c.get("now").has("offsetMillis") ? c.get("now").get("offsetMillis").asLong() : 0L;
            long want = c.get("value").asLong();
            out.add(DynamicTest.dynamicTest("valid/" + c.get("name").asText(), () ->
                    assertEquals(want, binder(ticking(instant, tick)).longValue("timestamp", Values.now(unit, offset)).longValue())));
        }
        for (JsonNode c : root.get("invalidArguments")) {
            String unit = c.get("unit").asText();
            long offset = c.get("offsetMillis").asLong();
            out.add(DynamicTest.dynamicTest("invalidArguments/" + c.get("name").asText(), () -> {
                ValueException e = assertThrows(ValueException.class,
                        () -> binder(ticking(instant, 0)).longValue("timestamp", Values.now(unit, offset)));
                assertEquals("timestamp", e.member());
            }));
        }
        out.add(DynamicTest.dynamicTest("one reading per call", () -> {
            LongSupplier clock = ticking(instant, tick);
            // The emitted request for the fixture's call: one binder, two
            // events, each timestamp bound in turn.
            Binder b = binder(clock);
            JsonNode sent = call.get("sent").get("logEvents");
            assertEquals(sent.get(0).get("timestamp").asLong(), b.longValue("logEvents", Values.now("epochMillis", -1L)).longValue());
            assertEquals(sent.get(1).get("timestamp").asLong(), b.longValue("logEvents", Values.now("epochMillis", 0L)).longValue());
            // The next call builds a fresh binder, and so reads the clock afresh.
            assertEquals(instant + tick, binder(clock).longValue("timestamp", Values.now("epochMillis", 0L)).longValue());
        }));
        return out;
    }

    private static File fixtureFile() {
        File dir = new File("").getAbsoluteFile();
        for (int i = 0; i < 8 && dir != null; i++) {
            File candidate = new File(dir, "compat/model/testdata/now/now.json");
            if (candidate.isFile()) {
                return candidate;
            }
            dir = dir.getParentFile();
        }
        return null;
    }
}

File: compat/suites/node-js-sdk/src/lib/scenario/nowfixture.test.ts
/**
 * The shared `$now` conformance fixture, compat/model/testdata/now.
 *
 * `$now` is the client's clock when a call is made, in epoch milliseconds,
 * plus an offset (compat/model/README.md § Values). This is where node-js-sdk
 * proves it against the same cases every other suite reads: each valid
 * spelling evaluates to the fixture's value with the clock pinned to its
 * instant, each invalid one is refused, the clock is read once per call
 * however many `$now`s the params hold, and a `$now` is never evaluated
 * outside a call's params.
 *
 * Run with: npm run test:unit
 */
import assert from "node:assert/strict";
import { readFileSync } from "node:fs";
import { describe, it } from "node:test";
import { fileURLToPath } from "node:url";

import { runScenarioTest, type ExecEnv } from "./executor.ts";
import {
  ExpressionError,
  checkNowArguments,
  evaluateParams,
  evaluateValue,
  type EvalContext,
} from "./expressions.ts";
import type { ScenarioTest, Value, ValueObject } from "./ir.ts";

/** Resolved from this module's own location, as blobfixture.test.ts does. */
const FIXTURE = fileURLToPath(
  new URL("../../../../../model/testdata/now/now.json", import.meta.url),
);

interface NowFixture {
  $comment: string;
  instant: number;
  valid: Array<{ name: string; now: unknown; value: number }>;
  invalid: Array<{ name: string; now: unknown }>;
  invalidArguments: Array<{ name: string; unit: string; offsetMillis: number }>;
  call: { tickMillis: number; params: ValueObject; sent: Record<string, unknown> };
}

function loadFixture(): NowFixture {
  const fixture = JSON.parse(readFileSync(FIXTURE, "utf8")) as NowFixture;
  const unknown = Object.keys(fixture).filter(
    (k) => !["$comment", "instant", "valid", "invalid", "invalidArguments", "call"].includes(k),
  );
  assert.deepEqual(unknown, [], `unknown keys in ${FIXTURE}`);
  assert.ok(
    fixture.valid.length > 0 && fixture.invalid.length > 0 && fixture.invalidArguments.length > 0,
    "the $now fixture may not be skipped by emptying it",
  );
  return fixture;
}

const fixture = loadFixture();

/** A clock that moves on every time it is read, from the fixture's instant. */
function tickingClock(): () => number {
  let next = fixture.instant;
  return () => {
    const reading = next;
    next += fixture.call.tickMillis;
    return reading;
  };
}

function ctx(extra: Partial<EvalContext> = {}): EvalContext {
  return { runId: "oc", group: "g", bag: new Map(), ...extra };
}

describe("the shared $now fixture", () => {
  for (const c of fixture.valid) {
    it(`valid/${c.name}`, () => {
      const got = evaluateValue({ $now: c.now } as Value, ctx({ nowMillis: fixture.instant }));
      assert.equal(got, c.value);
    });
  }
  for (const c of fixture.invalid) {
    it(`invalid/${c.name}`, () => {
      assert.throws(
        () => evaluateValue({ $now: c.now } as Value, ctx({ nowMillis: fixture.instant })),
        ExpressionError,
      );
    });
  }
  for (const c of fixture.invalidArguments) {
    it(`invalidArguments/${c.name}`, () => {
      assert.throws(() => checkNowArguments(c.unit, c.offsetMillis), ExpressionError);
    });
  }
  it("reads the clock once per call", () => {
    const sent = evaluateParams(fixture.call.params, ctx({ clock: tickingClock() }));
    assert.deepEqual(sent, fixture.call.sent);
  });
  it("never evaluates a $now outside a call's params", () => {
    assert.throws(
      () => evaluateValue({ $now: { unit: "epochMillis" } } as Value, ctx({ clock: tickingClock() })),
      ExpressionError,
    );
  });
});

describe("$now through the executor", () => {
  it("sends the client's clock as a number, one reading per call", async () => {
    const calls: Array<Record<string, unknown>> = [];
    const env: ExecEnv = {
      send: async (_op, params) => {
        calls.push(params);
        return { nextSequenceToken: "t" };
      },
      ctx: { runId: "oc", group: "logs-events", bag: new Map(), clock: tickingClock() },
      scenarioFile: "compat/model/authored/logs-events.json",
      log: () => {},
      sleep: async () => {},
    };
    const put: ScenarioTest = {
      name: "PutLogEvents",
      op: "PutLogEvents",
      call: { op: "PutLogEvents", params: fixture.call.params },
      assert: [{ kind: "responseField", checks: { "$.nextSequenceToken": { nonEmpty: true } } }],
    };
    await runScenarioTest(env, put);
    await runScenarioTest(env, put);
    assert.equal(calls.length, 2);
    assert.deepEqual(calls[0], fixture.call.sent);
    // The second call read the clock once more, a tick later.
    const second = calls[1].logEvents as Array<{ timestamp: number }>;
    assert.deepEqual(
      second.map((e) => e.timestamp),
      [fixture.instant + fixture.call.tickMillis - 1, fixture.instant + fixture.call.tickMillis],
    );
  });
});
